File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int start = 0;
        int maxLength = 0;
        int maxFreq = 0;

        vector<int> freq(26, 0);

        for(int end=0;end<s.size();end++)
        {
            freq[s[end] - 'A']++;

            maxFreq = max(maxFreq, freq[s[end] - 'A']);

            if( (end-start+1) - maxFreq > k)
            {
                freq[s[start]-'A']--;
                start++;
            }

            maxLength = max(maxLength, end-start+1);
        }

        return maxLength;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        Map<Character, Integer> mp = new HashMap<>();

        int start = 0;
        int freq = 0;
        int maxCount = 0;

        for(int end=0;end<s.length();end++)
        {
            char curr = s.charAt(end);

            mp.put(curr, 1+ mp.getOrDefault(curr, 0));

            freq = Math.max(freq, mp.get(curr));

            while ((end-start+1)-freq > k)
            {
                char startChar = s.charAt(start);

                mp.put(startChar, mp.get(startChar) - 1);
                start += 1;
            }

            maxCount = Math.max(maxCount, end-start+1);
        }

        return maxCount;
    }
}